python/graphstorm/dataloading/metadata.py

```diff
@@ -0,0 +1,954 @@
+        a unique edge type identifier.
+        """
+        etypes = []
+        for can_etype in self._etypes:
+            etypes.append(can_etype[1])
+
+        return etypes
+
+    def device(self):
+        """ Get the device where metadata graph is located.
+
+        """
+        return self._device
+
+    # Implementations of DGL specific APIs
+    @property
+    def canonical_etypes(self):
+        """ Get the canonical edge type list.
+
+        A canonical edge type is a three element tuple, where the first element is the source
+        node type, the second element is the edge type, and the third element is the destination
+        node type.
+        """
+        return self._etypes
+
+    @property
+    def nodes(self):
+        """ Provide an interface of metadata graph nodes
+
+        DGL's `nodes` property is a NodeView class for node data retrieval. This metadata DGL
+        graph cannot provide the NodeView class, but can provide its node type data retrieval
+        interface, i.e., g.nodes[ntype].data and g.nodes[ntype].data[nfeat_name].
+
+        Return
+        ------
+        dict : the dictionary of nodes and their DataView classes.
+        """
+
+        logging.warning(('This %s is a metadata graph that simulates a DGL graph without '
+            'actual graph structure, i.e., nodes and edges, but meta information, e.g., node '
+            'types, edge types, and node or edge feature names. You can use the '
+            '".edges[etype].data" or ".edges[etype].data[efeat_name]" interface '
+            'to retrieve edges\' data information.'),
+            self.__class__.__name__)
+        return self._nodes
+
+    @property
+    def edges(self):
+        """ Provide an interface of metadata graph edges
+
+        DGL's `edges` property is an EdgeView class for edge data retrieval. This metadata DGL
+        graph cannot provide the EdgeView class, but can provide its edge type data retrieval
+        interface, i.e., g.edges[etype].data and g.edges[etype].data[efeat_name].
+
+        Return
+        ------
+        dict : the dictionary of edges and their DataView classes.
+        """
+        logging.warning(('This %s is a metadata graph that simulates a DGL graph without '
+            'actual graph structure, i.e., nodes and edges, but meta information, e.g., node '
+            'types, edge types, and node or edge feature names. You can use the '
+            '".edges[etype].data" or ".edges[etype].data[efeat_name]" interface '
+            'to retrieve edges\' data information.'),
+            self.__class__.__name__)
+        return self._edges
+
+    def is_homogeneous(self):
+        """ Check if the metadata graph is a homogeneous one.
+        """
+        return self._graph_metadata.is_homo()
+
+    @property
+    def ndata(self):
+        """ Provide an interface for retrieving node data in a homogeneous metadata graph.
```

Comment:
What will happen if a user access ndata with a heterogeneous graph?

Reply:
Without knowledge of the consumers it's harder to evaluate this PR. If we control all instantiations of this object we can restrict its API to the minimum possible.

Is there a downstream situation where a consumer instantiates this object with a heterog graph? If yes, how does that consumer handle the case where it gets None as a return value?

For this kind of problem it's much easier to use type annotations (`def ndata(self) -> Optional[return_type]:`) . Consuming code will then be flagged if the caller tries to use the returned value without checking if it's None or not.

Reply:
> Although GS does not call this function, as a general API simulation, it is better to provide this API.

In this context who is instantiating this object? Its intended use is GraphStorm-internal right? Why not keep the API to the absolute minimum that GraphStorm needs to function?